src/headless/plugins/blocking/api.js

```diff
@@ -4,97 +4,103 @@ import { _converse, api, converse } from '@converse/headless/core.js';
-            const users_to_unblock = sizzle('item', stanza).map((item) => item.getAttribute('jid'));
-            users_to_unblock.forEach(_converse.blocked.get('set').delete, _converse.blocked.get('set'));
+    handleBlockingStanza ( stanza ) {
+        const action = stanza.firstElementChild.tagName;
+        const items = sizzle('item', stanza).map(item => item.getAttribute('jid'));
+        const msg_type = stanza.getAttribute('type');
+
+        log.debug(`handle blocking stanza Type ${msg_type} action ${action}`);
+        if (msg_type == 'result' && action == 'blocklist' ) {
+            log.debug(`resetting blocklist: ${items}`);
+            _converse.blocked.set({'set': new Set()});
+            items.forEach((item) => { _converse.blocked.get('set').add(item)});
+
+            /**
+            * Triggered once the _converse.blocked list has been fetched
+            * @event _converse#blockListFetched
+            * @example _converse.api.listen.on('blockListFetched', () => { ... });
+            */
+            api.trigger('blockListFetched', _converse.blocked.get('set'));
+            log.debug("triggered blockListFetched");
+
+        } else if (msg_type == 'set' && action == 'block') {
+            log.debug(`adding people to blocklist: ${items}`);
+            items.forEach((item) => { _converse.blocked.get('set').add(item)});
+            api.trigger('blockListUpdated', _converse.blocked.get('set'));
+        } else if (msg_type == 'set' && action == 'unblock') {
+            log.debug(`removing people from blocklist: ${items}`);
+            items.forEach((item) => { _converse.blocked.get('set').delete(item)});
+            api.trigger('blockListUpdated', _converse.blocked.get('set'));
         } else {
-            log.error('Received blocklist push update but could not interpret it.');
+            log.error("Received a blocklist push update but could not interpret it");
         }
-        // TODO: Fix this to not use the length as an update key, and
-        // use a more accurate update method, like a length-extendable hash
-        _converse.blocked.set({ 'len': _converse.blocked.get('set').size });
+        return true;
     },
 
     /**
-     * Blocks JIDs by sending an IQ stanza
-     * @method api.blockUser
-     *
-     * @param { Array } [jid_list] - The list of JIDs to block
+     * Send block/unblock IQ stanzas to the server for the JID specified
+     * @method api.sendBlockingStanza
+     * @param { String } action - "block", "unblock" or "blocklist"
+     * @param { String } iq_type - "get" or "set"
+     * @param { Array } [jid_list] - (optional) The list of JIDs to block or unblock
      */
-    async blockUser (jid_list) {
-        if (!_converse.disco_entities.get(_converse.domain)?.features?.findWhere({ 'var': Strophe.NS.BLOCKING })) {
-            return false;
-        }
+    async sendBlockingStanza ( action, iq_type = 'set', jid_list = [] ) {
```

Comment:
I think this function should be removed out of the `api.js` module and put inside `utils.js`.

It's a utility method used by API methods, but not an API method itself.

Reply:
Yes, any function that's not being exposed as part of the public API should be moved to `utils.js`.